argon2/src/lib.rs

```diff
@@ -0,0 +1,906 @@
+        } else {
+            digest.update(0u32.to_le_bytes());
+        }
+
+        digest.update(&(ad.len() as u32).to_le_bytes());
+        digest.update(ad);
+
+        digest.finalize()
+    }
+}
+
+/// Argon2 instance: memory pointer, number of passes, amount of memory, type,
+/// and derived values.
+///
+/// Used to evaluate the number and location of blocks to construct in each
+/// thread.
+pub struct Argon2_instance {
+    /// Memory blocks
+    memory: Vec<Block>,
+
+    /// Version
+    version: Argon2_version,
+
+    /// Number of passes
+    passes: u32,
+
+    /// Number of blocks in memory
+    memory_blocks: usize,
+
+    /// Segment length
+    segment_length: u32,
+
+    /// Lane length
+    lane_length: u32,
+
+    /// Number of lanes
+    lanes: u32,
+
+    /// Number of threads
+    threads: u32,
+
+    /// Argon2 type
+    argon2_type: Argon2_type,
+}
+
+impl Argon2_instance {
+    /// Function allocates memory, hashes the inputs with Blake, and creates first
+    /// two blocks.
+    ///
+    /// Returns struct containing main memory with 2 blocks per lane initialized.
+    fn initialize(
+        context: &Argon2_Context<'_>,
+        argon2_type: Argon2_type,
+        initial_hash: &digest::Output<Blake2b>,
+    ) -> Result<Self> {
+        // Align memory size
+        // Minimum memory_blocks = 8L blocks, where L is the number of lanes
+        let mut memory_blocks = context.m_cost;
+
+        if memory_blocks < 2 * ARGON2_SYNC_POINTS * context.lanes {
+            memory_blocks = 2 * ARGON2_SYNC_POINTS * context.lanes;
+        }
+
+        let segment_length = memory_blocks / (context.lanes * ARGON2_SYNC_POINTS);
+
+        // Ensure that all segments have equal length
+        let memory_blocks = (segment_length * context.lanes * ARGON2_SYNC_POINTS) as usize;
+        let memory = vec![Block::default(); memory_blocks];
+
+        let mut instance = Argon2_instance {
+            version: context.version,
+            memory,
+            passes: context.t_cost,
+            memory_blocks,
+            segment_length,
+            lane_length: segment_length * ARGON2_SYNC_POINTS,
+            lanes: context.lanes,
+            threads: context.threads,
+            argon2_type,
+        };
+
+        if instance.threads > instance.lanes {
+            instance.threads = instance.lanes;
+        }
+
+        // TODO(tarcieri): GENKAT
+        // initial_kat(initial_hash, context, instance->type);
+
+        // Creating first blocks, we always have at least two blocks in a slice
+        instance.fill_first_blocks(&initial_hash)?;
+
+        Ok(instance)
+    }
+
+    /// Function creates first 2 blocks per lane
+    fn fill_first_blocks(&mut self, blockhash: &[u8]) -> Result<()> {
+        let mut hash = [0u8; ARGON2_BLOCK_SIZE];
+
+        for l in 0..self.lanes {
+            // Make the first and second block in each lane as G(H0||0||i) or
+            // G(H0||1||i)
+            for i in 0u32..2u32 {
```

Comment:
Finally tracked down the bug in my original translation. This function needed to fill the first two blocks, but it was only filling one because I accidentally wrote the loop as `0u32..1u32` (which is non-inclusive).

With that corrected, it's passing an Argon2d test vector locally! 🎉